alg/gdal_point_templ.h

```diff
@@ -0,0 +1,263 @@
+/******************************************************************************
+ * $Id$
+ *
+ * Project:  GDAL Point abstraction
+ * Purpose:
+ * Author:   Javier Jimenez Shaw
+ *
+ ******************************************************************************
+ * Copyright (c) 2024, Javier Jimenez Shaw
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#ifndef GDAL_POINT_TEMPL_H_INCLUDED
+#define GDAL_POINT_TEMPL_H_INCLUDED
+
+/*! @cond Doxygen_Suppress */
+
+#include <array>
+#include <functional>
+#include <initializer_list>
+
+namespace gdal
+{
+template <typename T, std::size_t N> class RawPoint
+{
+  public:
+    using value_type = T;
+    using size_type = std::size_t;
+    using self_type = RawPoint<T, N>;
+
+    constexpr size_type size() const noexcept
+    {
+        return N;
+    }
+
+    RawPoint()
+    {
+    }
+
+    RawPoint(T x, T y)
```

Comment:
I'm not the best expert on advanced C++ features, but I guess you'd need something like a constructor with a variadic number of arguments: https://stackoverflow.com/questions/28866559/writing-variadic-template-constructor
And you probably want to ensure that the number of items passed is exactly N